### PR TITLE
snap-seccomp: add in-kernel bpf tests 

### PR DESCRIPTION
**Pull request by mvo5**
Based on https://github.com/snapcore/snapd/pull/3431 - this adds extra tests by generating test binaries and run them against the in-kernel bpf VM.

### CODE REVIEW

File: cmd/snap-seccomp/main_test.go [hunk 1]
```diff
@@ -129,6 +134,205 @@ func nativeEndian() binary.ByteOrder {
...
+}
+`)
+
+func lastKmsg() string {
+	output, err := exec.Command("dmesg").CombinedOutput()
+	if err != nil {
+		return err.Error()
+	}
+	l := strings.Split(string(output), "\n")
+	return l[len(l)-2]
```
Comment (jdstrand) [inline cmd/snap-seccomp/main_test.go]: This is racy and trying to pick a needle out of a potential kernel logging haystack. It might be better to use `lastKmsg()` only when you have an unexpected error and return something like:
```
Showing last 10 lines of dmesg:
show
your
last
10
lines
of
dmesg
output
right
here
```

File: cmd/snap-seccomp/main_test.go [hunk 2]
```diff
@@ -129,6 +134,205 @@ func nativeEndian() binary.ByteOrder {
...
+	// build syscall-runner helper
+	s.seccompSyscallRunner = filepath.Join(c.MkDir(), "seccomp_syscall_runner")
+	err = ioutil.WriteFile(s.seccompSyscallRunner+".c", seccompSyscallRunnerContent, 0644)
+	c.Assert(err, IsNil)
+	cmd = exec.Command("gcc", "-Werror", "-Wall", "-static", s.seccompSyscallRunner+".c", "-o", s.seccompSyscallRunner, "-Wl,-static", "-static-libgcc")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	c.Assert(err, IsNil)
+
```
Comment (jdstrand) [inline cmd/snap-seccomp/main_test.go]: Can you remove this newline?

File: cmd/snap-seccomp/main_test.go [hunk 3]
```diff
@@ -129,6 +134,205 @@ func nativeEndian() binary.ByteOrder {
...
+execve
+uname
+brk
+arch_prctl
+readlink
+access
+sysinfo
+exit
+# i386
+set_thread_area
```
Comment (jdstrand) [inline cmd/snap-seccomp/main_test.go]: You might need other on other architectures. If so, these can of easily be identified in the spread tests.

File: cmd/snap-seccomp/main_test.go [hunk 4]
```diff
@@ -129,6 +134,205 @@ func nativeEndian() binary.ByteOrder {
...
+	// syscallName;arch;arg1,arg2...
+	l := strings.Split(bpfInput, ";")
+	if len(l) > 1 && l[1] != "native" {
+		c.Logf("cannot use non-native in runBpfInKernel")
+		return
+	}
+	if strings.Contains(bpfInput, "PR_SET_ENDIAN") {
+		c.Logf("cannot run PR_SET_ENDIAN in runBpfInKernel, this actually switches the endianess and the program crashes")
+		return
+	}
```
Comment (jdstrand) [inline cmd/snap-seccomp/main_test.go]: Looking at `man syscall`, it seems we may also want to skip on the following syscalls: fadvise64_64, ftruncate64, posix_fadvise, pread64, pwrite64, readahead, sync_file_range, and truncate64. Please add a comment here that we just skip syscalls with architecture-specific details and in seccompSyscallRunnerContent (right above the call to `syscall()`) indicating that there might be architecture-specific requirements and to see `man syscall` for details.

File: cmd/snap-seccomp/main_test.go [hunk 5]
```diff
@@ -162,13 +366,8 @@ func simulateBpf(c *C, seccompWhitelist, bpfInput string, expected int) {
...
 	c.Assert(err, IsNil)
 	c.Check(out, Equals, expected, Commentf("unexpected result for %q (input %q), got %v expected %v", seccompWhitelist, bpfInput, out, expected))
-}
 
-func (s *snapSeccompSuite) SetUpSuite(c *C) {
-	// FIXME: we currently use a fork of x/net/bpf because of:
-	//   https://github.com/golang/go/issues/20556
-	// switch to x/net/bpf once we can simulate seccomp bpf there
-	bpf.VmEndianness = nativeEndian()
+	s.runBpfInKernel(c, seccompWhitelist, bpfInput, expected)
```
Comment (jdstrand) [inline cmd/snap-seccomp/main_test.go]: Now that you added this, can you adjust the comment above `simulateBpf` to have:
```
// simulateBpf first:
//  1. runs main.Compile() which will catch syntax errors and output to a file
//  2. takes the output file from main.Compile and loads it via
//     decodeBpfFromFile
//  3. parses the decoded bpf using the seccomp library and various
//     snapd functions
//  4. runs the parsed bpf through a bpf VM
//
// Then simulateBpf runs the policy through the kernel by calling
// runBpfInKernel() which:
//  1. runs main.Compile()
//  2. the program in seccompBpfLoaderContent with the output file as an
//     argument
//  3. the program in seccompBpfLoaderContent loads the output file BPF into
//     the kernel and executes the program in seccompBpfRunnerContent with the
//     syscall and arguments specified by the test
//
// In this manner, in addition to verifying policy syntax we are able to
// unit test the resulting bpf in several ways.
//
// Full testing of applied policy is done elsewhere via spread tests.
```

**Comment by mvo5** [comment]
Thanks @jdstrand ! I addressed all comments now.

File: cmd/snap-seccomp/main_test.go [hunk 6]
```diff
@@ -148,21 +153,232 @@ func nativeEndian() binary.ByteOrder {
...
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	c.Assert(err, IsNil)
+
+	// build syscall-runner helper
+	s.seccompSyscallRunner = filepath.Join(c.MkDir(), "seccomp_syscall_runner")
+	err = ioutil.WriteFile(s.seccompSyscallRunner+".c", seccompSyscallRunnerContent, 0644)
+	c.Assert(err, IsNil)
+	cmd = exec.Command("gcc", "-Werror", "-Wall", "-static", s.seccompSyscallRunner+".c", "-o", s.seccompSyscallRunner, "-Wl,-static", "-static-libgcc")
```
Comment (zyga) [inline cmd/snap-seccomp/main_test.go]: Does the `-static` matter here? I'm worried about another place that does static linking across the distro landscape.

File: cmd/snap-seccomp/main_test.go (hunk 3 above)
Comment (zyga) [inline cmd/snap-seccomp/main_test.go]: Yeah, I was about to write that this list is a great way to learn how linux userspace works across architectures. I don't mind having it but I bet this will bite us the first time this runs on s360 and you have to fix the tests.

File: cmd/snap-seccomp/main_test.go [hunk 7]
```diff
@@ -148,21 +153,232 @@ func nativeEndian() binary.ByteOrder {
...
+	// syscallName;arch;arg1,arg2...
+	l := strings.Split(bpfInput, ";")
+	if len(l) > 1 && l[1] != "native" {
+		c.Logf("cannot use non-native in runBpfInKernel")
+		return
+	}
+	// Note that we will need to also skip: fadvise64_64,
+	//   ftruncate64, posix_fadvise, pread64, pwrite64, readahead,
+	//   sync_file_range, and truncate64.
+	// Once we start using those. See `man syscall`
```
Comment (zyga) [inline cmd/snap-seccomp/main_test.go]: The comment and the next code block seem unrelated. Am I missing anything?

File: cmd/snap-seccomp/main_test.go [hunk 8]
```diff
@@ -148,21 +153,232 @@ func nativeEndian() binary.ByteOrder {
...
+
+	var syscallRunnerArgs [7]string
+	syscallNr, err := seccomp.GetSyscallFromName(l[0])
+	c.Assert(err, IsNil)
+	syscallRunnerArgs[0] = strconv.FormatInt(int64(syscallNr), 10)
+	if len(l) > 2 {
+		args := strings.Split(l[2], ",")
+		for i := range args {
+			// init with random number argument
+			syscallArg := (uint64)(rand.Uint32())
```
Comment (zyga) [inline cmd/snap-seccomp/main_test.go]: What's the point of this random value?

File: cmd/snap-seccomp/main_test.go [hunk 9]
```diff
@@ -233,7 +444,7 @@ func (s *snapSeccompSuite) TestCompile(c *C) {
 		{"read\nwrite\nexecve\n", "write", main.SeccompRetAllow},
 
 		// trivial denial
-		{"read", "execve", main.SeccompRetKill},
+		{"read", "ioctl", main.SeccompRetKill},
```
Comment (zyga) [inline cmd/snap-seccomp/main_test.go]: Why is this changed?

File: cmd/snap-seccomp/main_test.go [hunk 10]
```diff
@@ -449,10 +660,10 @@ func (s *snapSeccompSuite) TestRestrictionsWorkingArgsQuotactl(c *C) {
 		// good input
 		seccompWhitelist := fmt.Sprintf("quotactl %s", arg)
 		bpfInputGood := fmt.Sprintf("quotactl;native;%s", arg)
-		simulateBpf(c, seccompWhitelist, bpfInputGood, main.SeccompRetAllow)
+		s.simulateBpf(c, seccompWhitelist, bpfInputGood, main.SeccompRetAllow)
 		// bad input
-		for _, bad := range []string{"quotactl;native;99999", "read;native;"} {
```
Comment (zyga) [inline cmd/snap-seccomp/main_test.go]: Why is this changed?

File: cmd/snap-seccomp/main_test.go (hunk 6 above)
Comment (mvo5) [inline cmd/snap-seccomp/main_test.go]: Yes, it explodes with: `/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libc.a(libc-start.o): relocation R_X86_64_32 against undefined symbol _dl_starting_up' can not be used when making a shared object; recompile with -fPIC` without that.

File: cmd/snap-seccomp/main_test.go (hunk 7 above)
Comment (mvo5) [inline cmd/snap-seccomp/main_test.go]: I updated it a bit to clarify it some more. It is essentially about that we cannot test some syscalls using the in-kernel VM because when we run them things explodes.

File: cmd/snap-seccomp/main_test.go (hunk 8 above)
Comment (mvo5) [inline cmd/snap-seccomp/main_test.go]: Mostly to avoid that everything is set to "0" and we miss issues because we did not test other values.

File: cmd/snap-seccomp/main_test.go (hunk 9 above)
Comment (mvo5) [inline cmd/snap-seccomp/main_test.go]: Execve is now part of the `common` syscalls, i.e. it must be allowed so that we can actually run these tests here.

File: cmd/snap-seccomp/main_test.go [hunk 11]
```diff
@@ -148,21 +153,234 @@ func nativeEndian() binary.ByteOrder {
...
+}
+`)
+
+var seccompSyscallRunnerContent = []byte(`
+#define _GNU_SOURCE
+#include<unistd.h>
+#include<sys/syscall.h>
+#include<stdlib.h>
+int main(int argc, char **argv) {
+   int l[7];
```
Comment (niemeyer) [inline cmd/snap-seccomp/main_test.go]: The indentation here is different from the one above. We have three spaces here and a tab above, apparently.

**Comment by niemeyer** [comment]
@zyga This is ready for another pass.

File: cmd/snap-seccomp/main_test.go (hunk 6 above)
Comment (zyga) [reply cmd/snap-seccomp/main_test.go]: Aha, I don't know why this happens but it will be another thing that needs distro-patching as static linking to libseccomp is not a given. I'll investigate and see if I can make this nicer.

File: cmd/snap-seccomp/main_test.go (hunk 8 above)
Comment (zyga) [reply cmd/snap-seccomp/main_test.go]: Can you adjust the comment to explain this.

File: cmd/snap-seccomp/main_test.go (hunk 3 above)
Comment (jdstrand) [reply cmd/snap-seccomp/main_test.go]: That's ok, it's easy to fix. Note that this list is not very different from what was in the old snap-confine unit tests.

**Comment by jdstrand** [comment]
FYI, the changes since I approved this look fine (though I'm not sure we we chose 'webkit' for clang formatting, but not a blocker).

**Comment by mvo5** [comment]
@jdstrand I'm equally puzzled about the webkit formating  but I agree its not a blocker.

**Comment by mvo5** [comment]
@jdstrand I'm equally puzzled about the webkit formating  but I agree its not a blocker.

**Comment by niemeyer** [comment]
@jdstrand Can you please mark as an actual approval so it's more clear in the PR status?
